### DISCUSSION TITLE
Livewire and Cypress

### DISCUSSION THREAD

**Discussion by bhulsman** [livewire/livewire#3541] [1. Help] [answered]
I started using Cypress to get a testing suite that is usable no matter what frontend framework/library/tech I'm using. For example; I start with a basic form; just using a blade template and a controller, nothing fancy. But later, the form needs an input that has to many options for a select. Maybe I had 5 options to choose from, but with a 100 scrolling through the options becomes tedious. So I add a keyword field and let the user search for it.

Using Livewire that's pretty easy to do, but my default feature/controller test should now be a Livewire component test. I have to sack the original test and create a new one, when I'm just changing one input field on the whole form.

So... using Cypress I would add a testing suite on top of the whole app and I don't have to rewrite whole tests when a changing frontend tech.

But; I continue to get 419 errors from Livewire requests and I can't find any way to solve them, neither why the occur. I'm hoping someone here ran into this issue and solved it.

**Comment by ryatkins** [comment, marked as answer]
I started with Cypress, but quickly moved to Dusk so I had the best of both worlds, the power of PHPUnit and front-end testing.  Also, there are some nice Livewire macros in the core that help with the waiting of data to come back from the server.

I imagine there are better ways to login to your app via Cyrpess to not get the 419 error, but I had to manually login 

```
describe('Login', () => {
    it('Login to app', () => {
        cy.visit('http://app.test/login')

        cy.get('input[name="email"]').type('ryatkins@....')
        cy.get('input[name="password"]').type('password')
        cy.get('button').click()

       ....
```

On the other hand, with Dusk you can just use Laravel to login via `$browser->loginAs(User::find(4))`
```
class QuoteTest extends DuskTestCase
{
public function test_create_new_quote()
    {
        $this->browse(function ($browser) {

            $browser->loginAs(User::find(4))

                // New Quote
                // Find Customer
                ->visit('/new-quote/companies')
                ->assertSee('Customer Search')

                ->waitForLivewire()->type('companySearch', 'Apple')

...
```

In DuskTestCase to get Livewire macros:

```
abstract class DuskTestCase extends BaseTestCase
{
    use CreatesApplication;

    /**
     * Prepare for Dusk test execution.
     *
     * @beforeClass
     * @return void
     */
    public static function prepare()
    {

        Browser::mixin(new DuskBrowserMacros);

        if (! static::runningInSail()) {
            static::startChromeDriver();
        }

    }
```

**Reply by bhulsman** [reply]
Thanks! So it's worthwile to take Dusk for a spin too.

The https://github.com/laracasts/cypress solves some issues and besides the 419 errors it actually works well I should say, but I got to explore my options a bit ofcourse.

**Reply by joshhanley** [reply]
@bhulsman yeah personally I prefer Dusk over Cypress when testing Laravel stuff. Similar concept to Cypress but using PHPUnit.

It's also good because all my unit and integration tests can be run as the same time as my Dusk tests just by running PHPUnit.

And if you want some inspiration on how to use dusk, a lot of Livewire's code base is tested using dusk, so you can use that as examples of how to use it. Look in https://github.com/livewire/livewire/tree/master/tests/Browser

Hope this helps!

**Reply by bhulsman** [reply]
Awesome, thanks! ^^
